device/cuda/src/details/sparse_ccl.cu

```diff
@@ -0,0 +1,343 @@
+         * certain that this is necessary here, but the overhead is not that big
+         * and we might as well be safe rather than sorry.
+         */
+        __syncthreads();
+
+        /*
+         * This loop initializes the adjacency cache, which essentially
+         * translates the sparse CCL problem into a graph CCL problem which we
+         * can tackle with well-studied algorithms.
+         */
+        for (unsigned int tst = 0; tst * blockDim.x + threadIdx.x < size; ++tst) {
+            unsigned int tid = tst * blockDim.x + threadIdx.x;
+
+            unsigned int tc0 = cells[tid].channel0;
+            unsigned int tc1 = cells[tid].channel1;
+
+            /*
+             * First, we traverse the cells backwards, starting from the current
+             * cell and working back to the first, collecting adjacent cells
+             * along the way.
+             */
+            for (unsigned int j = tid - 1; j < tid; --j) {
+                /*
+                 * Since the data is sorted, we can assume that if we see a cell
+                 * sufficiently far away in both directions, it becomes
+                 * impossible for that cell to ever be adjacent to this one.
+                 * This is a small optimisation.
+                 */
+                if (
+                    cells[j].channel0 < tc0 - 1 &&
+                    cells[j].channel1 < tc1 - 1
+                ) {
+                    break;
+                }
+
+                /*
+                 * If the cell examined is adjacent to the current cell, save it
+                 * in the current cell's adjacency set.
+                 */
+                if (is_adjacent(cells[tid], cells[j])) {
+                    adjv[tst][adjc[tst]++] = j;
+                }
+            }
+
+            /*
+             * Now we examine all the cells past the current one, using almost
+             * the same logic as in the backwards pass.
+             */
+            for (unsigned int j = tid + 1; j < size; ++j) {
+                /*
+                 * Note that this check now looks in the opposite direction! An
+                 * important difference.
+                 */
+                if (
+                    cells[j].channel0 > tc0 + 1 &&
+                    cells[j].channel1 > tc1 + 1
+                ) {
+                    break;
+                }
+
+                if (is_adjacent(cells[tid], cells[j])) {
+                    adjv[tst][adjc[tst]++] = j;
+                }
+            }
+        }
+
+        /*
+         * Now that the data has initialized, we synchronize again before we
+         * move onto the actual processing part.
+         */
+        __syncthreads();
+
+        /*
+         * The algorithm finishes if an iteration leaves the arrays unchanged.
+         * This varible will be set if a change is made, and dictates if another
+         * loop is necessary.
+         */
+        bool gfc;
```

Comment:
Is this `bool` thread local?

Reply:
Yes! The sharing of whether any element of the union-find set was changed is shared between threads only at the end, when the `__any_sync` fires.